source/server-discovery-and-monitoring/server-monitoring.rst

```diff
@@ -742,6 +752,66 @@ The event API here is assumed to be like the standard `Python Event
           interrupt connection read
           close tempConnection
 
+RTT thread
+``````````
+
+The requirements in the `Measuring RTT`_ section can be statisfied with an
+addtional thread that periodically runs the isMaster command on a dedicated
+connection, for example:
+
+.. code-block:: python
+
+  class RttMonitor(Thread):
+    def __init__():
+        # Options:
+        serverAddress = serverAddress
+        connectTimeoutMS = connectTimeoutMS
+        heartbeatFrequencyMS = heartbeatFrequencyMS
+        # Internal state:
+        connection = Null
+        lock = Mutex()
+        movingAverage = MovingAverage()
+
+    def reset():
+        with lock:
+            movingAverage.reset()
+
+    def addSample(rtt):
+        with lock:
+            movingAverage.update(rtt)
+
+    def average():
+        with lock:
+            return movingAverage.get()
+
+    def run():
+        while this monitor is not stopped:
+            try:
+                rtt = pingServer()
+                addSample(rtt)
+            except (NetworkError, CommandError) as exc:
```

Comment:
Do we need to reset the average RTT here or is it enough for the server monitor to reset it when it errors?

Reply:
I don't think we should. If this thread calls reset() then it might lead to this race:
```
    def run():
....
          else:
              # The server does not support topologyVersion.
              set connection timeout to connectTimeoutMS
              start = time()
              response = call {isMaster: 1}
              rttMonitor.addSample(time()-start)
            # <--- RttMonitor calls reset() here. rttMonitor.average() is now 0.
            return ServerDescription(response, rtt=rttMonitor.average())
```

Reply:
Agreed. I asked because I talked to @jyemin about the Java code and he had structured it in a way where both the server and RTT monitors reset the RTT, so we weren't sure which one(s) should.